### ISSUE TITLE
Pretty-Print / Format JSON Response Bodies

### ISSUE THREAD

**Issue by Nezteb** [Orange-OpenSource/hurl#1760]
### Problem to solve

By default it seems that any JSON printed by `hurl` is not pretty-printed. Sometimes when I'm developing/debugging an API, I like to inspect the resulting JSON, which isn't doable in `hurl` without external tools.

This same feature request could also apply to HTML and XML, but at least for my use-case I am only asking for JSON. 😅 

### Proposal

Kind of similar to https://github.com/Orange-OpenSource/hurl/issues/1745, but I'd suggest either a CLI flag for this and/or a `[Option]` property to enable pretty printing for either an entire `.hurl` file or a single request.

Something like:
```
hurl test.hurl --pretty-print # or maybe something more descriptive and specific to JSON
```
and/or:
```
GET https://jsonplaceholder.typicode.com/users
[Option]
pretty-print: true # or maybe something more descriptive and specific to JSON
```

### Additional context and resources

I currently pipe my `hurl` commands to `jq` to get the pretty-printed version of the JSON: https://github.com/jqlang/jq

However, this only works if I'm only printing a single response body; otherwise the printed output isn't valid JSON.

**Comment by chrisguest75** [comment]
Or an option to dump the responses into folders under a subpath.  It would also help when writing tests and creating assertions. 

[Option]
responses-path: ./out/runs
write-responses: true

**Comment by jcamiel** [comment]
Hi @chrisguest75, unless I've mistaken your point, you can do this with [`--output`](https://hurl.dev/docs/manual.html#output):

```hurl
GET https://foo.com
[Options]
output: response.bin
HTTP 200
```

**Comment by chrisguest75** [comment]
Thanks, that does work for me.  I had to update to 4.3.0 first though.  Didn't realise I was using an older version.

**Comment by oatnog** [comment]
Although it would be neat to have prettier JSON handling built in, I've been achieving something similar by piping output to [jq](https://jqlang.org/).

**Comment by 1xX69** [comment]
I'm surprised and disappointed that this feature is still not supported

**Comment by jcamiel** [comment]
@Nezteb the issue is not open to contribution, but still open, we'll do it.

**Comment by jcamiel** [comment]
@lepapareil @fabricereix quick questions about the feature, I would like your opinions:

- I'm thinking about two more options: `--pretty` and `--no-pretty`. By default, and in interactive session, we'll prettify JSON ouput (`--pretty` is default):
    - If you do not want pretty output, you'll have to use `--no-pretty`
    - if you want pretty output while redirecting, you'll have to explicitely force pretty print with `--pretty`
- What about errors? Whether `--pretty` is implicit or set by user, what do we do if we have an error (JSON is not valid etc..)? I don't know how `httpie` is managing it for instance, I think we should silent fail if we have any error.
- When do we try to pretty print? Do we try do pretty print every response (whether or not the `Content-Type` response is JSON), or only if we detect JSON content type in the response headers? My opionion:
   - if `Content-Type` is present and of type JSON, try pretty print
   - if `Content-Type` is present and not of type JSON, do nothing (unless explicit `--pretty`)
   - if `Content-Type` is not present, try pretty print as best effort (silent failing if not valid JSON)
- `--pretty` take `--color`/ `--no-color` into account

**Comment by fabricereix** [comment]
Yes it makes sense to have a consistent behavior between color and pretty (implicit w/o option and explicit with the options).

Relating to an invalid JSON, we can indeed leave the response as received without producing any errors. That's also the behavior of httpie.

For the implicit mode, I think I would prefer to activate the pretty formatting only with the content-type header.
If the response is a valid JSON, but with the wrong content-type header, I would leave the response as received.
But this is not the behavior of httpie.

        $ curl localhost:8000/sample.txt
        {"name":"Bob"}
        
        $ n/http GET localhost:8000/sample.txt
        HTTP/1.0 200 OK
        Content-Length: 15
        Content-type: text/plain
        Date: Thu, 21 Aug 2025 10:34:07 GMT
        Last-Modified: Thu, 21 Aug 2025 10:33:39 GMT
        Server: SimpleHTTP/0.6 Python/3.13.1
        
        {
            "name": "Bob"
        }

**Comment by jcamiel** [comment]
With "implicit" prettify, we prettify the body if the content is JSON and do nothing otherwise. We'll have a small penalty for trying to parse JSON in anycase. Maybe we can try prettify if Content-Type is explicitely JSON and text like?

Or maybe simpler:
- without `--pretty`, pretiffy only JSON-like content type
- with `--pretty`, try to pretify no matter what the content type is

**Comment by lepapareil** [comment]
@jcamiel @fabricereix, not sure yet on the best solution, we should map out all real use cases including interactive/non-interactive mode and future features like websocket to clarify the default option and how smart it should be. Better to discuss this together live 🧐.

**Comment by jcamiel** [comment]
@fabricereix @lepapareil Now that we have an "easy" way of pretty printing JSON, I see two places where we could also use it(or not):


- `--error-format`: with `--error-format long` we output the HTTP response whenever there are asserts failure
- `--very-verbose`: in this mode, we also log the HTTP response.

In details:

1. `--error-format`

With Hurl 7.0.0:

<img width="1938" height="1014" alt="Image" src="https://github.com/user-attachments/assets/e328f589-ffa4-4ba2-8025-a092855cc33f" />

With Hurl 7.1.0:

<img width="1906" height="2146" alt="Image" src="https://github.com/user-attachments/assets/fe9a0061-b4a1-4df0-8af9-971c0d7015e3" />

I like the "visual" improvment (a lot of time with Hurl 7.0.0, I copy/paste the body response, prettify it to make the assert fix easy).

2. `--very-verbose`

With Hurl 7.0.0:

<img width="1928" height="2240" alt="Image" src="https://github.com/user-attachments/assets/df802067-f152-406d-b7fb-6dd5e498fb8b" />

With Hurl 7.1.0:

<img width="1350" height="2402" alt="Image" src="https://github.com/user-attachments/assets/1646627d-4167-40c7-8bfe-c0a2b31b1d3d" />

With Hurl 7.1.0 variant with no color:

<img width="1440" height="2594" alt="Image" src="https://github.com/user-attachments/assets/e130dc54-9429-403d-8637-b5495206dfc4" />


Maybe in this debug mode we should indicate that the output has been prettified

Opinions?

**Comment by fabricereix** [comment]
I wouldn't expect any prettified in the log, just raw stuff.

**Comment by jcamiel** [comment]
Ok for the log, and for `--error-format long`? We're displaying colored response headers and the HTTP response as well
